aws/resource_aws_budget.go

```diff
@@ -0,0 +1,303 @@
+	return map[string]*schema.Schema{
+		"name": {
+			Type:          schema.TypeString,
+			Optional:      true,
+			ConflictsWith: []string{"name_prefix"},
+		},
+		"name_prefix": {
+			Type:     schema.TypeString,
+			Optional: true,
+			ForceNew: true,
+		},
+		"budget_type": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"limit_amount": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"limit_unit": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"include_credit": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"include_other_subscription": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"include_recurring": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"include_refund": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"include_subscription": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"include_support": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"include_tax": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"include_upfront": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  true,
+		},
+		"use_blended": {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+		"time_period_start": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"time_period_end": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Default:  "2087-06-15_12:00",
+		},
+		"time_unit": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"cost_filters": &schema.Schema{
+			Type:     schema.TypeMap,
+			Optional: true,
+			Computed: true,
+		},
+	}
+}
+
+func resourceAwsBudgetCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AWSClient).budgetconn
+	accountID := meta.(*AWSClient).accountid
+	budget, err := newBudget(d)
+	if err != nil {
+		return fmt.Errorf("failed creating budget: %v", err)
+	}
+
+	createBudgetInput := new(budgets.CreateBudgetInput)
+	createBudgetInput.SetAccountId(accountID)
+	createBudgetInput.SetBudget(budget)
+	_, err = client.CreateBudget(createBudgetInput)
+	if err != nil {
+		return fmt.Errorf("create budget failed: %v", err)
+	}
+
+	d.SetId(*budget.BudgetName)
+	return resourceAwsBudgetUpdate(d, meta)
+}
+
+func resourceAwsBudgetRead(d *schema.ResourceData, meta interface{}) error {
+	budgetName := d.Id()
+	describeBudgetOutput, err := describeBudget(budgetName, meta)
+	if isBudgetNotFoundException(err) {
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("describe budget failed: %v", err)
+	}
+
+	if _, ok := d.GetOk("name"); ok {
+		d.Set("name", describeBudgetOutput.Budget.BudgetName)
+	}
+
+	d.Set("budget_type", describeBudgetOutput.Budget.BudgetType)
+	d.Set("limit_amount", describeBudgetOutput.Budget.BudgetLimit.Amount)
```

Comment:
These nested attributes could be very panic prone. They should be checked, e.g.

```go
budgetLimit := describeBudgetOutput.Budget.BudgetLimit
if budgetLimit != nil {
  d.Set("limit_amount", budgetLimit.Amount)
}
```

Although ideally, this is another case where we should introduce nested attributes and the state set becomes:

```go
if err := d.Set("limit", flattenBudgetsBudgetLimit(describeBudgetOutput.Budget.BudgetLimit)); err != nil {
  return err
}
```